### PR TITLE
Fix @betterC testsuite

### PR DESCRIPTION
**Pull request by wilzbach**
Found in https://github.com/dlang/phobos/pull/7408

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -10924,6 +10924,7 @@ if (isInputRange!Range && !isInstanceOf!(SortedRange, Range))
     }
 
     ///
+    static if (is(Range : int[]))
```
Comment (n8sh) [inline std/range/package.d]: Did this get included by accident?

Comment (wilzbach) [reply std/range/package.d]: No, these tests are now also run and included. We can either limit the effect or move the tests out of the struct. This is exactly why a `version(StdUnittest)` would be nice.

File: std/typecons.d [hunk 2]
```diff
@@ -1184,7 +1184,7 @@ if (distinctFieldNames!(Specs))
             Returns:
                 A `size_t` representing the hash of this `Tuple`.
          */
-        size_t toHash() const nothrow @safe
+        size_t toHash()() const nothrow @safe
```
Comment (n8sh) [inline std/typecons.d]:
```suggestion
        size_t toHash() const nothrow @safe
```

And below replace
```d
static if (__traits(compiles, h = .hashOf(field[i])))
    const k = .hashOf(field[i]);
else
    // Workaround for when .hashOf is not both @safe and nothrow.
    // BUG: Improperly casts away `shared`!
    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```

with something like

```d
static if (__traits(compiles, h = .hashOf(field[i])))
    const k = .hashOf(field[i]);
else static if (is(T : shared U, U) && __traits(compiles, h = .hashOf((() @trusted => cast(U) field[i])())))
    // BUG: Improperly casts away `shared`!
    const k = .hashOf((() @trusted => cast(U) field[i])());
else
    // Workaround for when .hashOf is not both @safe and nothrow.
    // BUG: Improperly casts away `shared`!
    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```

This change is reasonable because we also have to cast away shared to call `typeid(T).getHash`. So if removing shared would also let us call `.hashOf` there is no reason to involve `typeid`. This is enough to make the current tests compile and pass both with and without -betterC.

**Comment by n8sh** [comment]
Pushing a change to make the tests pass.

File: std/typecons.d [hunk 3]
```diff
@@ -1192,9 +1192,16 @@ if (distinctFieldNames!(Specs))
...
                     const k = .hashOf(field[i]);
                 else
+                {
                     // Workaround for when .hashOf is not both @safe and nothrow.
-                    // BUG: Improperly casts away `shared`!
-                    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
+                    static if (is(T : shared U, U) && __traits(compiles, (U* a) nothrow @safe => .hashOf(*a))
+                            && !__traits(hasMember, T, "toHash"))
+                        // BUG: Improperly casts away `shared`!
+                        const k = .hashOf(*(() @trusted => cast(U*) &field[i])());
```
Comment (atilaneves) [inline std/typecons.d]: Why is this needed?

Comment (n8sh) [reply std/typecons.d]: So it compiles in betterC. When not in betterC it isn't expected to have different result than using typeid.

**Comment by PetarKirov** [comment]
This is wrong, see e857ce1127b5ff0a3b2b54159d86ff31c00dc0dd
